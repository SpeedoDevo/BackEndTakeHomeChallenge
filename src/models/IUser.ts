import IUserModel from './IUserModel'

export default interface IUser {
  getId(): number
  getName(): string
  getUsername(): string
  getAvatar(): string | undefined
  updateName(name: string): this
  updateAvatar(avatar?: string): this
  toObject(): IUserModel
}
