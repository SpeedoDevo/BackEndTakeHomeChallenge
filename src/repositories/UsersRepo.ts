import IUser from '../models/IUser'
import User from '../models/User'

const users = [
  {
    id: 1,
    name: 'Daenerys Targaryen',
    username: 'thequeenbee',
  },
  {
    id: 2,
    name: 'Jon Snow',
    username: 'forthebrotherhood'
  },
  {
    id: 3,
    name: 'Sansa Stark',
    username: 'joffery.fan'
  },
  {
    id: 4,
    name: 'Ned Stark',
    username: 'honouraboveall'
  },
  {
    id: 5,
    name: 'Cersei Lannister',
    username: 'bestmuminwesteros'
  }
]

export const getLastIdUsers = () => users.reduce((prev, current) => (prev.id > current.id) ? prev : current).id
export const getAllUsers = () => users.map(user => new User(user))
export const findUserById = (id: number) => users.find(user => user.id === id)
export const addUser = (user: IUser) => users.push({
  id: user.getId(),
  name: user.getName(),
  username: user.getUsername(),
})
