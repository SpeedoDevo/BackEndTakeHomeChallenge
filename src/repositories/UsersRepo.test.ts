import IUser from '../models/IUser'
import { addUser, getAllUsers, getLastIdUsers, getUserById } from './UsersRepo'

describe('UsersRepo', () => {
  describe('getLastIdUsers', () => {
    it('should return the last id', () => {
      expect(getLastIdUsers()).toBe(5)
    })
  })

  describe('getAllUsers', () => {
    it('should return all users', () => {
      expect(getAllUsers()).toMatchSnapshot()
    })
  })

  describe('addUser', () => {
    beforeEach(() => {
      addUser({
        getId: () => 6,
        getName: () => 'new user',
        getUsername: () => 'admin',
        getAvatar: () => 'undefined',
      } as IUser)
    })

    it('should add the new user', () => {
      expect(getAllUsers()).toMatchSnapshot()
    })
  })

  describe('getUserById', () => {
    it('should get the user', () => {
      expect(getUserById(5)).toMatchSnapshot()
    })

    it('should give null if not found', () => {
      expect(getUserById(666)).toBeNull()
    })
  })
})
