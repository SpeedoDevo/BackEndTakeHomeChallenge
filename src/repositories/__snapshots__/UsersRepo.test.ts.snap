// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersRepo addUser should add the new user 1`] = `
Array [
  User {
    "avatar": undefined,
    "id": 1,
    "name": "Daenerys Targaryen",
    "username": "thequeenbee",
  },
  User {
    "avatar": undefined,
    "id": 2,
    "name": "Jon Snow",
    "username": "forthebrotherhood",
  },
  User {
    "avatar": undefined,
    "id": 3,
    "name": "Sansa Stark",
    "username": "joffery.fan",
  },
  User {
    "avatar": undefined,
    "id": 4,
    "name": "Ned Stark",
    "username": "honouraboveall",
  },
  User {
    "avatar": undefined,
    "id": 5,
    "name": "Cersei Lannister",
    "username": "bestmuminwesteros",
  },
  User {
    "avatar": "undefined",
    "id": 6,
    "name": "new user",
    "username": "admin",
  },
]
`;

exports[`UsersRepo getAllUsers should return all users 1`] = `
Array [
  User {
    "avatar": undefined,
    "id": 1,
    "name": "Daenerys Targaryen",
    "username": "thequeenbee",
  },
  User {
    "avatar": undefined,
    "id": 2,
    "name": "Jon Snow",
    "username": "forthebrotherhood",
  },
  User {
    "avatar": undefined,
    "id": 3,
    "name": "Sansa Stark",
    "username": "joffery.fan",
  },
  User {
    "avatar": undefined,
    "id": 4,
    "name": "Ned Stark",
    "username": "honouraboveall",
  },
  User {
    "avatar": undefined,
    "id": 5,
    "name": "Cersei Lannister",
    "username": "bestmuminwesteros",
  },
]
`;

exports[`UsersRepo getUserById should get the user 1`] = `
User {
  "avatar": undefined,
  "id": 5,
  "name": "Cersei Lannister",
  "username": "bestmuminwesteros",
}
`;
